db/sqldb/analysis.go

```diff
@@ -9,25 +9,41 @@ import (
 	"github.com/jackc/pgx/v5/pgtype"
 )
 
-const analyseSelectColumns = `
-	analysis.id,
-	analysis.analysis_type,
-	analysis.owner_id,
-	analysis.pacta_version_id,
-	analysis.portfolio_snapshot_id,
-	analysis.name,
-	analysis.description,
-	analysis.created_at,
-	analysis.ran_at,
-	analysis.completed_at,
-	analysis.failure_code,
-	analysis.failure_message`
+func analysisQuery(where string) string {
+	return fmt.Sprintf(`
+		WITH selected_analysis_ids AS (
+			SELECT id FROM analysis %[1]s
+		)
+		SELECT
+			analysis.id,
+			analysis.analysis_type,
+			analysis.owner_id,
+			analysis.pacta_version_id,
+			analysis.portfolio_snapshot_id,
+			analysis.name,
+			analysis.description,
+			analysis.created_at,
+			analysis.ran_at,
+			analysis.completed_at,
+			analysis.failure_code,
+			analysis.failure_message,
+			aas.analysis_artifact_ids
+		FROM
+			analysis
+			LEFT JOIN (
+				SELECT 
+					analysis_id,
+					ARRAY_AGG(analysis_artifact.id) AS analysis_artifact_ids
+				FROM analysis_artifact 
+				WHERE analysis_id IN (SELECT id FROM selected_analysis_ids)
+				GROUP BY analysis_id
+			) aas ON aas.analysis_id = analysis.id
+		%[1]s;
+`, where)
```

Comment:
Same here, this seems like lots of nested subqueries that we don't need. What's wrong with something like:

```suggestion
	return fmt.Sprintf(`
		SELECT
			analysis.id,
			analysis.analysis_type,
			analysis.owner_id,
			analysis.pacta_version_id,
			analysis.portfolio_snapshot_id,
			analysis.name,
			analysis.description,
			analysis.created_at,
			analysis.ran_at,
			analysis.completed_at,
			analysis.failure_code,
			analysis.failure_message,
			ARRAY_AGG(analysis_artifact.id)
		FROM analysis
		LEFT JOIN analysis_artifact
			ON analysis_artifact.analysis_id = analysis.id
		%[1]s ;; This is the WHERE clause I think. Also, I'd put the `WHERE ` here, so that all the callers don't have to remember it
		GROUP by analysis.id;
	`, where)
```

Reply:
This is mostly for consistency with the other query structure. The other query looked JUST like you're proposing. But then, when I added a second nested relationship table to the query, it broke. The advantage I see of structuring it this way is to prime it so that additional joins will follow the same structure, and be less likely to fall into the pitfall I spent a few hours having to debug.

Reply:
Copying this here for the discussion: https://github.com/RMI-PACTA/app/pull/91#discussion_r1437297560